Add options dialog to Reflectometry GUI [PART 3]

giovannidisiena opened the pull request:
**Description of work.**
This PR is **3** of **6**.

The warning for discarding changes has been implemented.

**To test:**
Complete a code review and check that the warning is displayed when expected (e.g. editing and closing a batch, loading a new batch on top of existing changes, etc) and the negative case (e.g. warning should not be displayed if a batch is first saved, then closed/overwritten). Full manual testing instructions will follow when subsequent PRs are merged.

Partially resolves #26533

*This does not require release notes* because **they will be added in a subsequent PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

giovannidisiena committed:
Integrate OptionsDialog into MainWindow

Re #26533

giovannidisiena committed:
Implement WarnDiscardChanges

Re #26533

giovannidisiena committed:
Re #26533 Add tests and mock objects

giovannidisiena committed:
Allow model to be passed to presenter to fix OptionsDialogPresenterTest

And other changes to avoid ctest warnings Re #26533

giovannidisiena committed:
Add unused macro to close event

... to prevent compiler warning Re #26533

giovannidisiena committed:
Fix code changed by file handler refactoring

Re #26533

giovannidisiena committed:
Fix typo, correct header guard and remove unnecessary declaration

Re #26533

giovannidisiena committed:
Address review comments with renaming and minor refactoring

Re #26533

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.h:
```diff
@@ -57,6 +57,15 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL MainWindowPresenter
+  bool isCloseEventPrevented() override;
+  bool isCloseBatchPrevented(int batchIndex) const override;
+  bool isOperationPrevented() const override;
+  bool isOperationPrevented(int tabIndex) const override;
```
Can this be `private`? Would a more specific name be better, say, something like `isOverwritingBatchAllowed`?

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.h:
```diff
@@ -57,6 +57,15 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL MainWindowPresenter
+  bool isOperationPrevented() const override;
+  bool isOperationPrevented(int tabIndex) const override;
+  bool isBatchUnsaved(int batchIndex) const override;
+  bool isAnyBatchUnsaved() override;
```
private?

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.cpp:
```diff
@@ -161,8 +160,73 @@ bool MainWindowPresenter::isAnyBatchAutoreducing() const {
+
+/**
+     Checks whether there are unsaved changes in any batch
+     * @return : Bool on whether there are changes
```
Very minor point but the formatting of this comment is a bit weird!

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Save/QtSaveView.h:
```diff
@@ -102,6 +102,13 @@ public slots:
+  void connectSettingsChange(QComboBox &edit);
+  void connectSettingsChange(QCheckBox &edit);
+  void connectSettingsChange(QRadioButton &edit);
+  void connectInstrumentSettingsWidgets();
```
Should this be named `connectSaveSettingsWidgets`?

Also, it does not appear to be called - I think this is just an omission, because the warnings aren't working at the moment.

giovannidisiena replied:
Yes, it should be, though I'm not sure where it should get called?

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Save/SavePresenter.cpp:
```diff
@@ -42,6 +42,10 @@ void SavePresenter::acceptMainPresenter(IBatchPresenter *mainPresenter) {
 }
 
+void SavePresenter::notifySettingsChanged() {
+  m_mainPresenter->notifySettingsChanged();
```
I don't think we can call this here because I think it will end up resetting the state of rows in the table. The Save tab is a bit different to the other tabs, because if we're just changing something like the filter criteria in the search box, this doesn't affect the reduction configuration, so therefore shouldn't reset the rows' state. However, I don't think this is called at the moment anyway so this bug is not showing up!

gemmaguest reviewed (changes requested):
Works well in these scenarios:

- Warning appears when loading a batch after table has changed, and respects the setting.
- Warning appears when starting autoprocess, and respects the setting.
- Warning does not appear after pausing and restarting autoprocessing even if the table has changed, because the table does not get overwritten unless the search criteria have changed. The warning does appear if the search criteria have changed.
- Warning appears after changing settings on Event/Experiment/Instrument tabs.

However, seems to be incorrect in these scenarios:

- Warning does not appear after editing the Save tab.
- Warning when starting autoprocess does not respect the setting.
- Batch is incorrectly marked as unsaved after processing a row (when state has changed but no content in the table has actually changed)
- Warning incorrectly appears when I `Restore Defaults` in Experiment and Instrument tabs if nothing has changed. I don't think it's necessary to have a warning here so we can just remove this.

giovannidisiena commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.h:
```diff
@@ -57,6 +57,15 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL MainWindowPresenter
+  bool isCloseBatchPrevented(int batchIndex) const override;
+  bool isOperationPrevented() const override;
+  bool isOperationPrevented(int tabIndex) const override;
+  bool isBatchUnsaved(int batchIndex) const override;
```
Which in particular? I added the methods here as they are `IMainWindow` overrides but I think the existing methods were public. Should they all be private?

giovannidisiena commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.h:
```diff
@@ -92,6 +101,7 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL MainWindowPresenter
   std::unique_ptr<MantidWidgets::ISlitCalculator> m_slitCalculator;
   std::unique_ptr<IOptionsDialogPresenter> m_optionsDialogPresenter;
   std::unique_ptr<IBatchPresenterFactory> m_batchPresenterFactory;
+  bool m_isUnsaved;
```
Yes, in terms of displaying discard changed warnings this is done from the batches. I think this is left over from the thought that we could/should ask to save the state of the GUI on closing/closing Mantid, so there would be a method to iterate through all the batch flags and set this `m_isUnsaved` to true if any of those are true. Anyhow, I will remove it for now.

giovannidisiena commented on qt/scientific_interfaces/ISISReflectometry/GUI/MainWindow/MainWindowPresenter.h:
```diff
@@ -57,6 +57,15 @@ class MANTIDQT_ISISREFLECTOMETRY_DLL MainWindowPresenter
+  bool isWarnDiscardChangesChecked() const override;
+  bool isCloseEventPrevented() override;
+  bool isCloseBatchPrevented(int batchIndex) const override;
+  bool isOperationPrevented() const override;
```
Which are you referring to?

giovannidisiena replied:
I think it is still used via the `RunsPresenter`

giovannidisiena committed:
Fix warnings behaving incorrectly

Now appears after editing save tab
Doesn't appear for autoprocessing unless changes will be lost
Batch not marked as unsaved after processing
Removed warning on restore defaults

giovannidisiena committed:
Merge branch '26533_refl_add_options_gui_1' into 26533_refl_add_options_gui_3

giovannidisiena committed:
Reorder BatchPresenter initialiser list

gemmaguest commented:
@giovannidisiena I'm getting a segfault just opening the reflectometry interface. Have you seen this?

giovannidisiena commented:
> @giovannidisiena I'm getting a segfault just opening the reflectometry interface. Have you seen this?

@gemmaguest no, that's strange, it was fine when I tested it... The only thing I can think of is maybe moving some of the calls in `7a2006b3926cfca86d947d97a70897b44b16e3a7` but I haven't observed it.

gemmaguest commented:
> > @giovannidisiena I'm getting a segfault just opening the reflectometry interface. Have you seen this?
> 
> @gemmaguest no, that's strange, it was fine when I tested it... The only thing I can think of is maybe moving some of the calls in commit `7a2006b3926cfca86d947d97a70897b44b16e3a7` but I haven't observed it.

Ok I'll try rebuilding in case I messed something up

gemmaguest commented:
The crash is happening in `SavePresenter::notifySettingsChanged` because `m_mainPresenter` is not initialized. If you initialize it to `nullptr` in the constructor then you should see the crash as well. It's currently not initialized at all, which probably explains why you're getting different behaviour. This should have been done in the original code but was missing. Apologies - you would have seen the crash if it had been done!

giovannidisiena commented:
@gemmaguest I have initialised `m_mainPresenter` but still not seeing the crash. There is a call to `m_savePresenter->acceptMainPresenter` in the `BatchPresenter` constructor which seems to me would behave as expected. `notifySettingsChanged` is not called until after the GUI is constructed, only when the user changes the state of one of the widgets. There are some calls in the `savePresenter` constructor which would result in the notification being given so I think the connection in the view would need to be done after this. There is not currently a `notifyee` for the save tab so I am going to create one.

giovannidisiena committed:
Connect saveSettingsWidgets after other constructor calls

Re #26533

gemmaguest commented:
> @gemmaguest I have initialised `m_mainPresenter` but still not seeing the crash. There are some calls in the `savePresenter` constructor which would result in the notification being given before it has accepted the main presenter so I think the connection in the view would need to be done after this. Since I can't recreate the crash you will probably need to verify that this solves things!

Yes, that's fixed it :+1:

giovannidisiena commented:
@gemmaguest would I be able to get a re-review please? The test failure will be fixed once I rebase onto master, though there doesn't seem much point to do this for any other than the base branch.

gemmaguest reviewed (approved):
Code looks good now and all the cases I mentioned before work correctly :+1: :+1:
